cpp/ql/src/semmle/code/cpp/models/implementations/SmartPointer.qll

```diff
@@ -18,20 +18,20 @@ private class UniqueOrSharedPtr extends Class, PointerWrapper {
 }
 
 /** Any function that unwraps a pointer wrapper class to reveal the underlying pointer. */
-private class PointerWrapperDataFlow extends DataFlowFunction {
-  PointerWrapperDataFlow() {
+private class PointerWrapperFlow extends TaintFunction, DataFlowFunction {
+  PointerWrapperFlow() {
     this = any(PointerWrapper wrapper).getAnUnwrapperFunction() and
     not this.getUnspecifiedType() instanceof ReferenceType
   }
 
-  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
-    input.isQualifierAddress() and output.isReturnValue()
-    or
-    input.isQualifierObject() and output.isReturnValueDeref()
-    or
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
     input.isReturnValueDeref() and
     output.isQualifierObject()
   }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and output.isReturnValue()
```

Comment:
After we started seeing FPs like https://github.com/github/codeql/issues/5709, I'm getting nervous about rules like this one, pretending a smart pointer is the same thing as the pointer it contains. But for the purpose of this PR, at least it doesn't seem more risky than what we had before.

Reply:
Yeah, I agree. Hopefully I'll have an idea for how to fix this later today and we can discuss it on Zoom tomorrow for our 1-on-1.